### PR TITLE
Increase the versions of python we test against on CI and tox. Remove…

### PR DESCRIPTION
**Pull request by mbasset**
## Tracker Ticket

https://quandl.atlassian.net/browse/AP-1631

## Description

* Remove the `.get` from model calls.
* Increase the # of versions of python we test against

## Dependency

* None

### CODE REVIEW

File: quandl/model/data.py [hunk 1]
```diff
@@ -9,10 +9,10 @@ class Data(DataListOperation, DataMixin, ModelBase):
     def __init__(self, data, **options):
         self.meta = options['meta']
         converted_column_names = list([Util.methodize(x) for x in options['meta']['column_names']])
-        self.raw_data = Util.convert_to_dates(OrderedDict(list(zip(converted_column_names, data))))
+        self._raw_data = Util.convert_to_dates(OrderedDict(list(zip(converted_column_names, data))))
 
     def __getattr__(self, k):
-        if k[0] == '_':
+        if k[0] == '_' and k != '_raw_data':
```
Comment (ccleung) [inline quandl/model/data.py]: why is this needed?

File: quandl/model/model_base.py [hunk 2]
```diff
@@ -1,21 +1,26 @@
...
+        return self.__get_raw_data__()
 
     def __getattr__(self, k):
         if k[0] == '_':
             raise AttributeError(k)
-        elif k in self.raw_data:
-            return self.raw_data[k]
+        elif k in self.__get_raw_data__():
+            return self.__get_raw_data__()[k]
         raise AttributeError(k)
```
Comment (ccleung) [inline quandl/model/model_base.py]: just to confirm, this will still error out if the attribute does not exist? e.g., if we did `quandl.Database('NSE').foo` it will error out?

File: quandl/model/model_list.py [hunk 3]
```diff
@@ -1,7 +1,10 @@
 class ModelList(object):
     def __init__(self, klass, values, meta):
         self.klass = klass
-        self.values = list([klass(x, meta=meta) for x in values])
+        if hasattr(klass, 'get_code_from_meta'):
+            self.values = list([klass(klass.get_code_from_meta(x), x, meta=meta) for x in values])
```
Comment (ccleung) [inline quandl/model/model_list.py]: why is this needed? for merged_datasets?

File: quandl/model/dataset.py [hunk 4]
```diff
@@ -12,12 +12,25 @@ class Dataset(GetOperation, ListOperation, ModelBase):
...
         return "%s/metadata" % cls.default_path()
 
+    @classmethod
+    def get_code_from_meta(cls, metadata):
+        return "%s/%s" % (metadata['database_code'], metadata['dataset_code'])
+
+    def __init__(self, code, raw_data=None, **options):
+        ModelBase.__init__(self, code, raw_data)
+        self.database_code = self.code.split("/")[0]
+        self.dataset_code = self.code.split("/")[1]
```
Comment (ccleung) [inline quandl/model/dataset.py]: do we need to protect against bad inputs? e.g., `CODENAME` will cause a crash here

File: quandl/model/model_base.py [hunk 5]
```diff
@@ -1,21 +1,26 @@
 class ModelBase(object):
-    def __init__(self, raw_data, **options):
-        self.raw_data = raw_data
+    def __init__(self, code, raw_data=None, **options):
+        self.code = code
+        self._raw_data = raw_data
```
Comment (ccleung) [inline quandl/model/model_base.py]: can we turn this back to `raw_data`? I think before users could easily access the dictionary of the response via `Object.raw_data`

File: quandl/model/merged_dataset.py [hunk 6]
```diff
@@ -9,71 +9,25 @@
...
-
-    @classmethod
-    def _get_request_params(cls, dataset_code):
-        if isinstance(dataset_code, tuple):
-            return dataset_code[1]
-        return {}
+    def __init__(self, dataset_codes, **options):
+        self.dataset_codes = dataset_codes
+        self._datasets = None
+        self.raw_data = None
```
Comment (ccleung) [inline quandl/model/merged_dataset.py]: what's going on here? can't we just re-uase ModelBase's `_raw_data` ? (or if we change back `ModelBase` to use `raw_data` again, nothing needs to change here)

File: quandl/model/merged_dataset.py [hunk 7]
```diff
@@ -154,8 +103,76 @@ def _in_query_param(self, name, **options):
...
+        # raw data will be initialized internally
+        self._initialize_raw_data(self._datasets)
+
+        return self._datasets
+
+    def _initialize_raw_data(self, datasets):
+        self._raw_data = {}
+        if not datasets:
+            return self._raw_data
+        self.raw_data = datasets[0].__get_raw_data__().copy()
```
Comment (ccleung) [inline quandl/model/merged_dataset.py]: we are mixing `_raw_data` and `raw_data`, what's the difference between the two?

File: quandl/model/model_base.py [hunk 8]
```diff
@@ -1,21 +1,26 @@
...
 
     def __getitem__(self, k):
-        return self.raw_data[k]
+        return self.__get_raw_data__()[k]
 
     def data_fields(self):
-        return list(self.raw_data.keys())
+        return list(self.__get_raw_data__().keys())
+
+    def __get_raw_data__(self):
```
Comment (ccleung) [inline quandl/model/model_base.py]: this seems to be defined twice? once here, and once in `GetOperation.py`. It looks like our models will inherit from both `GetOperation` and `ModelBase`, we should eliminate one of these?

Comment (ccleung) [inline quandl/model/model_base.py]: Edit: I see how this might be done on purpose - if GetOperation is inherited from, then we would have the http request GET operation, otherwise we would just simply return `_raw_data`?

File: quandl/model/data.py (hunk 1 above)
Comment (mbasset) [inline quandl/model/data.py]: went and threw and error because i renamed `raw_data` to `_raw_data` to indicate its private now. In which case it would match k[0] == '_' and it would throw an error.

File: quandl/model/model_base.py (hunk 2 above)
Comment (mbasset) [inline quandl/model/model_base.py]: I need to double check but it should react the same as it did before and raise an error I believe.

File: quandl/model/model_list.py (hunk 3 above)
Comment (mbasset) [inline quandl/model/model_list.py]: Correct!

File: quandl/model/dataset.py (hunk 4 above)
Comment (mbasset) [inline quandl/model/dataset.py]: Thought about that. I am not sure what the right direction to go is. We could throw a custom error in this case. Do you have something in mind?

File: quandl/model/model_base.py (hunk 5 above)
Comment (mbasset) [inline quandl/model/model_base.py]: The reason I renamed it to `_raw_data` is cause they should really be calling `__get_raw_data__()` method to get the raw data. Even in this case it's not a great idea to call the directly however if we want to expose it that would be the one to rename.

File: quandl/model/merged_dataset.py (hunk 6 above)
Comment (mbasset) [inline quandl/model/merged_dataset.py]: Raw data works differently now for modelbase than for merged dataset. That why I changed it to no longer use the super method. I need to rename this to _raw_data however. That is incorrect atm.

File: quandl/model/merged_dataset.py (hunk 7 above)
Comment (mbasset) [inline quandl/model/merged_dataset.py]: They should both be `_raw_data` now.

File: quandl/model/model_base.py (hunk 8 above)
Comment (mbasset) [inline quandl/model/model_base.py]: Correct! Yes its duplicated for that exact reason.

**Comment by mbasset** [comment]
@ccleung made a few adjustments and responded to comments.

File: quandl/model/dataset.py (hunk 4 above)
Comment (ccleung) [inline quandl/model/dataset.py]: `BadRequestError` and maybe an example of the valid format? or point to the docs.

Comment (ccleung) [inline quandl/model/dataset.py]: `BadRequestError` and maybe an example of the valid format? or point to the docs.

**Comment by ccleung** [comment]
:+1: see what you can do with the parsing of the `database/dataset_code`

File: quandl/model/dataset.py (hunk 4 above)
Comment (mbasset) [inline quandl/model/dataset.py]: Ya i shall raise some meaningful error.
